lesson 5/OpenCVtest/tensor.cpp

```diff
@@ -0,0 +1,504 @@
+		}
+		return tmp;
+	}
+	else {
+		throw WrongSize();
+	}
+}
+///////////////// TENSOR 3 D /////////////////////////////////////
+
+Tensor3D::Tensor3D(std::initializer_list<Tensor2D> data_) {
+	auto it = data_.begin();
+	auto in_shape = it->getShape();
+	for (std::size_t i = 0; i < data_.size() - 1; ++i) {
+		if (in_shape != it->getShape()) { throw WrongShapeTensorException(); }
+	}
+	std::copy(data_.begin(), data_.end(), std::back_inserter(data));
+	shape = { data.size(), in_shape[0], in_shape[1] };
+}
+
+void Tensor3D::printElements() {
+	for (const auto& cur : data) {
+		cur.printElements();
+		std::cout << std::endl;
+	}
+}
+
+void swap(Tensor3D& left, Tensor3D& right) {
+	std::swap(left.data, right.data);
+	std::swap(left.shape, right.shape);
+}
+
+Tensor3D::Tensor3D(const Tensor3D& copy) {
+	data = copy.data;
+	shape = copy.shape;
+}
+
+Tensor3D::Tensor3D(Tensor3D&& move) {
+	swap(*this, move);
+}
+
+Tensor3D& Tensor3D::operator=(const Tensor3D& copyAss) {
+	Tensor3D tmp(copyAss);
+	swap(*this, tmp);
+	return *this;
+}
+
+Tensor3D& Tensor3D::operator=(Tensor3D&& moveAss) {
+	Tensor3D tmp(std::move(moveAss));
+	swap(*this, tmp);
+	return *this;
+}
+
+std::array<std::size_t, 3> Tensor3D::getShape() {
+	return shape;
+}
+
+void Tensor3D::transpose() {
+	for (auto& cur : data) { cur.transpose(); }
+}
+
+std::vector<double> Tensor3D::flatten() const {
+	std::vector<double> flattend(shape[0] * shape[1] * shape[2]);
+	std::size_t i = 0;
+	for (auto& cur : data) {
+		std::vector<double> tmp = cur.flatten();
+		std::copy(tmp.begin(), tmp.end(), flattend.begin() + i++ * shape[1] * shape[2]);
+	}
+	return flattend;
+}
+
+Tensor3D::Tensor3D(const cv::Mat& mat) {
+	const int channels = mat.channels();
+	switch (channels) {
+		case 1: {
+			Tensor2D tmp_mono(mat);
+			data.push_back(tmp_mono);
+			shape = { (std::size_t)channels, (std::size_t)mat.rows, (std::size_t)mat.cols };
+			break;
+		}
+		case 3: {
+			cv::Mat chans[3];
+			cv::split(mat, chans);
+			Tensor2D tmp_B(chans[0]), tmp_G(chans[1]), tmp_R(chans[2]);
+			data = { tmp_B, tmp_G, tmp_R };
+			shape = { (std::size_t)channels, (std::size_t)mat.rows, (std::size_t)mat.cols };
+			break;
+		}
+		default: {
+			throw WrongSize();
+		}
+	}
+}
+void Tensor3D::pushChannel(const Tensor2D& channel) {
+	data.push_back(channel);
+	++shape[0];
+}
+double& Tensor3D::at(std::size_t channel, std::size_t row, std::size_t column) {
+	if (channel >= shape[0] || channel < 0 || row >= shape[1] || row < 0 || column >= shape[1] || column < 0) {
+		throw WrongSize();
+	}
+	//std::cout << "{{{  AT }}}" << std::endl;
+	return data[channel].getValue(row, column);
+}
+
+Tensor3D::operator cv::Mat() {
+	switch (shape[0]) {
+		case 1: {
+			cv::Mat tmp = fillDataInMat(shape[0]);
+			return tmp;
+			break;
+		}
+		case 3: {
+			cv::Mat mat;
+			std::vector<cv::Mat> channels;
+			for (int channel = 0; channel < shape[0]; ++channel) {
+				cv::Mat tmp = fillDataInMat(channel);
+				channels.push_back(tmp);
```

Comment:
since `tmp` is a temporary and you don't need it after this line, it'd better to  use move constructor of `cv::Mat`, not copy